### PR TITLE
feat(*): add ID's to targets

### PR DESCRIPTION
**Pull request by gcottrell1**
- adds id's for the targets, these will be used for the conditionals in linking to access the confidence levels instead of the camelCased meaning/intent as numbers at the start were causing issues.
- uses a simple function to generate a random 5 character string for the id if one is not passed in as part of the options.

let me know if the id generator is shit 😆

### CODE REVIEW

File: lib/target/Intent.ts [hunk 1]
```diff
@@ -8,10 +9,15 @@ export { IOptions };
 export class Intent implements IMutable {
   readonly type: Type.Intent = Type.Intent;
   intent: string;
+  id: string;
 
   constructor(init: IOptions | IImmutable | string) {
     const { intent } = typeof init === 'string' ? { intent: init } : init;
+    const id = typeof init === 'string' || !init.id ? camelCase(intent) : init.id;
```
Comment (mattyclarkson) [inline lib/target/Intent.ts]: Pull the `id` out in the destructuring:

```ts
const { intent, id } = typeof init === 'string' ? { intent: init } : init;
this.intent = intent;
this.id = id === undefined ? camelCase(this.intent) : id;
```

File: lib/target/camelCase.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+const camelCaseRegex = /(\S+)\s*/g;
+
+const sanitizeRegex = /[^a-zA-Z]/g;
+
+function capitalize(s: string): string {
+  return `${s.slice(0, 1).toUpperCase()}${s.slice(1).toLowerCase()}`;
+}
+
+export default function camelCase(value: string): string {
```
Comment (mattyclarkson) [inline lib/target/camelCase.ts]: I usually export both as a module export and default:

```ts
export function camelCase(value: string): string {
}

export default camelCase;
```

Do we want to export this in `index.ts`?

File: lib/target/Intent.ts (hunk 1 above)
Comment (mattyclarkson) [reply lib/target/Intent.ts]: Might wanna check both properties:

```ts
if (!this.intent) {
  throw new TypeError(`Target intent must not be empty`);
} else if (!this.id) {
  throw new TypeError(`Target ID for '${this.intent}' intent must not be empty`);
}
```

File: lib/target/Intent.ts [hunk 3]
```diff
@@ -8,10 +9,15 @@ export { IOptions };
...
+    // tslint:disable-next-line:max-line-length
+    if (id === '') { throw new TypeError(`Id cannot be blank, if the intent begins with a non-letter character, you must set the id manually`); }
     this.intent = intent;
+    this.id = id;
 
     const descriptors = Object.getOwnPropertyDescriptors(this);
     descriptors.type.writable = false;
```
Comment (mattyclarkson) [inline lib/target/Intent.ts]: Add the descriptors for `id` (`configurable = false`)

File: lib/target/Meaning.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import { IImmutable, IMutable, IOptions, IStatic } from '@target/IMeaning';
 import Type from '@target/Type';
+import camelCase from './camelCase';
```
Comment (mattyclarkson) [inline lib/target/Meaning.ts]: Can import this with an `@target` if you want to.

File: lib/target/camelCase.ts (hunk 2 above)
Comment (mattyclarkson) [reply lib/target/camelCase.ts]: Might actually be worth writing some unit tests for this function? Are these the expected results?

  * `123abc-yuh` -> `abcYuh`?
  * `aty123abc-yuh` -> `aty123AbcYuh`?
  * `__aty123__abc-yuh` -> `aty123AbcYuh`?

Comment (mattyclarkson) [reply lib/target/camelCase.ts]: You **might** want to change this function to `id` and export it as `targetId`. Then if we change our mind in the future to do something other than `camelCase` the function won't have a wonky name...

**Comment by gcottrell1** [comment]
@mattyclarkson should be good to review again now

File: lib/index.ts [hunk 5]
```diff
@@ -1,3 +1,7 @@
+// id
+import id from '@target/id';
+export { id as targetId };
```
Comment (mattyclarkson) [inline lib/index.ts]: Combine this into one export:

```ts
export { id as targetId } from '@target/id';
```

File: lib/target/IIntent.ts [hunk 6]
```diff
@@ -11,6 +11,7 @@ export interface IImmutable extends ITargetImmutable<Type.Intent> {
    * The actual target
    */
   readonly intent: string;
+
```
Comment (mattyclarkson) [inline lib/target/IIntent.ts]: Remove unneeded new line

File: lib/target/IMeaning.ts [hunk 7]
```diff
@@ -21,13 +21,15 @@ export interface IMutable extends IImmutable, ITargetMutable<Type.Meaning> {
    * The mutable meaning target
    */
   meaning: string;
+
```
Comment (mattyclarkson) [inline lib/target/IMeaning.ts]: Remove unneeded new line

File: lib/target/Intent.ts [hunk 8]
```diff
@@ -8,15 +9,32 @@ export { IOptions };
 export class Intent implements IMutable {
   readonly type: Type.Intent = Type.Intent;
   intent: string;
+  id: string;
 
   constructor(init: IOptions | IImmutable | string) {
-    const { intent } = typeof init === 'string' ? { intent: init } : init;
+    const { intent, id } = typeof init === 'string' ? { intent: init, id: undefined } : init;
```
Comment (mattyclarkson) [inline lib/target/Intent.ts]: You can do default arguments for destructuring:

```ts
const { intent, id = undefined } = typeof init === 'string' ? { intent: init } : init;
```

**Pull request merged**